server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java

```diff
@@ -37,57 +37,44 @@
-
-        private final Map<Class<? extends MetadataFieldMapper>, MetadataFieldMapper> metadataMappers = new LinkedHashMap<>();
+    public static final class Builder {
+        private final Map<Class<? extends MetadataFieldMapper>, MetadataFieldMapper> metadataMappers;
         private final RootObjectMapper rootObjectMapper;
         private final Mapper.BuilderContext builderContext;
         private final IndexSettings indexSettings;
         private final IndexAnalyzers indexAnalyzers;
-        private final DocumentMapperParser documentMapperParser;
         private final DocumentParser documentParser;
 
         private Map<String, Object> meta;
 
         public Builder(RootObjectMapper.Builder builder, MapperService mapperService) {
-            this.indexSettings = mapperService.getIndexSettings();
-            this.indexAnalyzers = mapperService.getIndexAnalyzers();
-            this.documentMapperParser = mapperService.documentMapperParser();
-            this.documentParser = mapperService.documentParser();
+            this(builder, mapperService.getIndexSettings(), mapperService.getIndexAnalyzers(), mapperService.documentParser(),
+                mapperService.getMetadataMappers());
+        }
+
+        Builder(RootObjectMapper.Builder builder,
+                IndexSettings indexSettings,
+                IndexAnalyzers indexAnalyzers,
+                DocumentParser documentParser,
+                Map<Class<? extends MetadataFieldMapper>, MetadataFieldMapper> metadataMappers) {
+            this.indexSettings = indexSettings;
+            this.indexAnalyzers = indexAnalyzers;
+            this.documentParser = documentParser;
             this.builderContext = new Mapper.BuilderContext(indexSettings.getSettings(), new ContentPath(1));
             this.rootObjectMapper = builder.build(builderContext);
-
-            final DocumentMapper existingMapper = mapperService.documentMapper();
-            final Map<String, TypeParser> metadataMapperParsers =
-                mapperService.mapperRegistry.getMetadataMapperParsers(indexSettings.getIndexVersionCreated());
-            for (Map.Entry<String, MetadataFieldMapper.TypeParser> entry : metadataMapperParsers.entrySet()) {
-                final String name = entry.getKey();
-                final MetadataFieldMapper existingMetadataMapper = existingMapper == null
-                        ? null
-                        : (MetadataFieldMapper) existingMapper.mappers().getMapper(name);
-                final MetadataFieldMapper metadataMapper;
-                if (existingMetadataMapper == null) {
-                    final TypeParser parser = entry.getValue();
-                    metadataMapper = parser.getDefault(mapperService.documentMapperParser().parserContext());
-                } else {
-                    metadataMapper = existingMetadataMapper;
-                }
-                metadataMappers.put(metadataMapper.getClass(), metadataMapper);
-            }
```

Comment:
it is not fantastic that this is now moved to MapperService. On the other hand this code block required DocumentMapper as well as DocumentMapperParser (which is also one of the callers) and a bunch of other components that are easily available in MapperService and need to be carried around otherwise through DocumentMapperParser. So this is an ok compromise I thought. We may want to think about further cleanups around how we share the most common objects.

Reply:
I wonder if we can make this neater by changing the signature of MetadataFieldMapper.getDefault()? Looking at the various implementations, we only need `indexVersionCreated()` and `isIdFieldDataEnabled()`, which we can abstract into a MetadataParseContext interface - that removes the reliance on DocumentMapperParser at least.

Reply:
I tried this. Given the current state of things, I could replace the call to DocumentMapperParser with a Supplier<ParserContext>. And even switching to a smaller context, the supplier would still be needed because idFieldDataEnabled is not carried anywhere. Moving back this code where it used to be, I would have to carry around the following: 
- Map<String, MetadataFieldMapper.TypeParser> metadataMapperParsers
- Supplier<Mapper.TypeParser.ParserContext> parserContextSupplier
- Function<String, Mapper> mapperLookup

especially the last one is odd, but the best I came up to not do this is what I did in this PR, expose a method from mapper service. What do you think?

Reply:
This is definitely better.  I'm not completely happy with the metadata field builder abstractions, but maybe that will become clearer as we go on.

Reply:
with the latest update, the third argument to lookup mappers would no longer be needed. Do you want to me to try and put this method back where it was, or are you ok with it being in MapperService?

Reply:
Let's leave it in MapperService for now, we can always move it later if it gets too annoying.